sbtc/src/leb128.rs

```diff
@@ -0,0 +1,556 @@
+        let mut cursor = Cursor::new(&buffer[..]);
+
+        cursor.set_position(pos);
+        cursor.read_leb128()
+    }
+}
+
+#[cfg(test)]
+mod proptests {
+    use super::*;
+    use proptest::prelude::*;
+
+    // Constants for proptest strategies
+    const MAX_PROPTEST_ITERATIONS: u32 = 10000;
+
+    // Value range constants for stratified testing. We use separate tests
+    // for these ranges as proptest uses pseudo-random sampling and otherwise
+    // may not get enough coverage for the different value ranges.
+    const VALUE_MAX_1_BYTE: u64 = (1 << 7) - 1; // 127
+    const VALUE_MAX_2_BYTES: u64 = (1 << 14) - 1; // 16,383
+    const VALUE_MAX_3_BYTES: u64 = (1 << 21) - 1; // 2,097,151
+    const VALUE_MAX_4_BYTES: u64 = (1 << 28) - 1; // 268,435,455
+    const VALUE_MAX_5_BYTES: u64 = (1 << 35) - 1; // 34,359,738,367
+    const VALUE_MAX_6_BYTES: u64 = (1 << 42) - 1; // 4,398,046,511,103
+    const VALUE_MAX_7_BYTES: u64 = (1 << 49) - 1; // 562,949,953,421,311
+    const VALUE_MAX_8_BYTES: u64 = (1 << 56) - 1; // 72,057,594,037,927,935
+    const VALUE_MAX_9_BYTES: u64 = (1 << 63) - 1; // 9,223,372,036,854,775,807
+    const VALUE_MAX_10_BYTES: u64 = u64::MAX; // 18,446,744,073,709,551,615
+
+    /// Number of payload bits per LEB128 byte (excluding continuation bit)
+    const BITS_PER_BYTE: u32 = 7;
+
+    /// Returns maximum value encodable in n bytes
+    /// Used to identify size boundaries for encoding decisions
+    const fn max_value_for_bytes(n: u32) -> u64 {
+        (1u64 << (BITS_PER_BYTE * n)) - 1
+    }
+
+    /// Returns minimum value requiring exactly n+1 bytes
+    /// Used to identify size transition points
+    const fn min_value_for_next_bytes(n: u32) -> u64 {
+        1u64 << (BITS_PER_BYTE * n)
+    }
+
+    /// Helper function for property-based testing of LEB128 encoding
+    ///
+    /// Performs comprehensive validation of a single value:
+    /// - Verifies the value encodes to exactly the expected number of bytes
+    /// - Confirms that size calculation matches actual encoded size
+    /// - Ensures perfect round-trip encoding/decoding
+    ///
+    /// This function is central to validating both the correctness and
+    /// compression efficiency of the LEB128 implementation across the
+    /// full range of possible values.
+    ///
+    /// ## Parameters
+    /// * `value` - The integer value to test
+    /// * `expected_len` - The expected encoded size in bytes
+    fn test_value(value: u64, expected_len: usize) -> Result<(), TestCaseError> {
+        let mut bytes = Vec::new();
+        Leb128::encode_into(value, &mut bytes);
+
+        prop_assert_eq!(
+            bytes.len(),
+            expected_len,
+            "Value {} should encode to {} bytes",
+            value,
+            expected_len
+        );
+        prop_assert_eq!(Leb128::calculate_size(value), expected_len);
+
+        let (decoded, _) = Leb128::try_decode(&bytes).unwrap();
+        prop_assert_eq!(decoded, value);
```

Comment:
nit: we could also check that bytes read matches `expected_len`

Reply:
Good call, [9e11140](https://github.com/stacks-network/sbtc/pull/1523/commits/9e111402f9cc91a0c87cb6bd58d9699950fba1b0)